### ISSUE TITLE
prange with random / shuffled iteration

### ISSUE THREAD

**Issue by songololo** [numba/numba#7060]
## Feature request

`prange` currently makes it possible to explicitly parallelise a loop (assuming `parallel=True`).

There are situations where iteration over a range should ideally be randomised / shuffled so that the CPU load can be spread more evenly across all threads.

In my case I am using `prange` to iterate nodes in a graph, and regions of the graph entail heavier computation than other regions in the graph. This results in certain threads (spawned via `prange`) finishing sooner than other threads, which will then block and complete more slowly than otherwise possible.

This could be alleviated by randomising the iteration over range.

**Comment by esc** [comment]
@songololo thank you for raising this, I have marked it as a feature request. One question: would it be possible to `shufffle` the iterator so as to randomize the task list ahead of time?

**Comment by songololo** [comment]
@esc yes, if a shuffled array can be passed to the iterator that would work fine. Thanks.

**Comment by stuartarchibald** [comment]
I think the issue here that as a `prange` induced loop execution scheduled isn't load balanced it doesn't perform so well in the case of an uneven work distribution? i.e. the number of iterations is equally distributed across the thread-count in equal size "chunks" such that if the work isn't evenly distributed in the iteration space threads are likely to end up waiting. This PR https://github.com/numba/numba/pull/6025 (once merged) might help as the chunksize can be set to e.g. 1 (or small) so that each thread only does one iteration and then has to request the next iteration from the scheduler.

**Comment by songololo** [comment]
@stuartarchibald yes, this sounds like it would work.
